File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // iterarive solution using a stack, O(n) time and space
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> in_order;
        stack<TreeNode*> s;
        TreeNode* curr = root;

        while (curr || !s.empty()) {
            while (curr) {
                s.push(curr);
                curr = curr->left;
            }
            curr = s.top();
            s.pop();
            in_order.push_back(curr->val);
            curr = curr->right;
        }
        return in_order;
    }
};

File: Data Structures & Algorithms/range-sum-of-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # iterative DFS solution, O(n) time and space
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        stack = [root]
        curr_sum = 0

        while stack:
            curr_node = stack.pop()
            if not curr_node:
                continue
            
            if low <= curr_node.val <= high:
                curr_sum += curr_node.val
            
            if curr_node.val > low:
                stack.append(curr_node.left)
            if curr_node.val < high:
                stack.append(curr_node.right)

        return curr_sum
        